enhancements/ingress/ingress-dashboard.md

```diff
@@ -0,0 +1,199 @@
+
+## Motivation
+
+While ingress related metrics already exist and are accessible in the OpenShift Console (via the menu “Observe” > Metrics”), there is no consolidated view presenting a summary of them. There are existing dashboards in other areas (such as etcd, compute resources, etc.), but networking today is less represented there, despite its importance for monitoring and troubleshooting.
+
+In addition, product management has shown interest in providing and making more visible some cluster-wide statistics such as the number of routes and shards in use.
+
+This [request for enhancement](https://issues.redhat.com/browse/NE-1059) is also considered here, although it will not be entirely fulfilled as we don't intend here to create any new metric.
+
+### Goals
+
+1. Review and select a subset of the available metrics that should be included in the new dashboard. This should ideally be done with the help of the NetEdge team expertise and/or SREs.
+2. Design a new dashboard with these metrics.
+3. Update the Ingress Operator to deploy this dashboard, making it accessible for Cluster Monitoring consumption.
+
+### Non-Goals
+
+- This work is not intended to provide a comprehensive set of metrics all at once. Instead, the intent is to "start small", setting in place all the mechanisms in code, and iterate later based on feedback to add or amend elements in the dashboard without any code change.
+- This enhancement does not include any new metric creation or exposition: only already available metrics are considered. If discussions lead to consider the creation of new metrics, this could be the purpose of a follow-up enhancement.
+
+### User Stories
+
+> As a cluster administrator, I want to get a quick overview of general cluster statistics, such as the number of routes or shards in use.
+> As a cluster administrator, I want to get a quick insight in incoming traffic statistics, such as on latency and HTTP errors.
+
+## Proposal
+
+To make a new dashboard discoverable by Cluster Monitoring, a `ConfigMap` needs to be created in the namespace `openshift-config-managed`, containing a static dashboard definition in Grafana format (JSON). The dashboard datasource has to be Cluster Monitoring's Prometheus.
+
+The Ingress Operator is responsible for creating and reconciling this `ConfigMap`. We assume all metrics used in the dashboard are present unconditionally, which allows us to create a static dashboard unconditionally as well.
+In the event where conditional display should be introduced later (e.g. an alternative to HAProxy being implemented, resulting is different metrics to display), different approaches can be discussed: the dashboard could be dynamically amended via JSON manipulation; or several dashboards could be embedded and selectively installed.
+
+The content of the dashboard will be further discussed and refined, including during review and verification.
```

Comment:
I expect that all of the dashboards implemented by your team will follow a common template that allows users to drill down on a namespace, time span, and other basic grouping details.  For Ingress, we would also like to allow a user to drill down on the Ingress controller instance they're interested in.  This way, if user wants to go from a top level system view of the last 24 hours, to a specific controller and namespace in the last hour, for example, that they can easily do this.

If you have this settled, presenting it here would be great.

Reply:
We don't have a fully free hand on what we can do here, as we're only supplying the dashboard data, and all the dashboard mechanics belongs to the console. Much of what you wrote seems doable, with some gotchas:

- Time selection is implemented - nothing special to do about that

- Grouping (e.g. by ingress controller instance) is possible, but is something pre-defined per chart, in the promQL query. E.g. we can have a chart for total MBps received per ingress controller instance, but the user can't change it to show a different grouping. We might create more charts if we want to show more groupings. If the users want more customized queries, they can click an "Inspect" link on any chart, and modify the query from there.

- Filtering (e.g. by namespace) is possible, but this is a dashboard global setting, meaning that it will impact all charts in the dashboard, and for that reason it could be problematic to show charts that wouldn't be affected by this filter if their metrics don't have the corresponding label. For example if we have a Namespace filter at the top of the dashboard, but some of the charts showing non-namespaced metrics (e.g. an `haproxy_frontend_something` metric that doesn't relate to routes / namespaces), that would be a design flaw and could turn out to be misleading for the users who could make wrong assumptions about the data they're seeing. If that becomes a problem because we want a filter X but some metrics don't have that label, the typical solution is to split into more dashboards, so that each dashboard has filters relevant for all its charts. For this EP we target a single dashboard, so we might be limited in terms of filtering. It could be for a follow-up improvement. Here again, when users want more customization, they can still click the "inspect" link and write their own queries.

- About the ingress controller instance (@frobware mentioned that as well): we checked and it's present indeed for some metrics, which means we can definitely show some charts with grouping by ingress controller instance.